### ISSUE TITLE
API implementation via C or via Python?

### ISSUE THREAD

**Issue by ajschumacher** [rstudio/tensorflow#105]
Hi! [Kevin](https://blog.rstudio.org/author/kevinusheyrstudio/) suggested [on the RStudio board](https://support.rstudio.com/hc/en-us/community/posts/115005611307) that I ask about this here: I’m looking at the variety of APIs out now for TensorFlow, and wondering about this one, and how it relates to the main TensorFlow project. It seems like the TensorFlow project wants people to code to the C API, whereas here it looks like you're wrapping the Python API, if I understand correctly. Are there any plans to change the approach? Will RStudio’s package continue to follow changes to the Python API, etc.? What are your thoughts? Thanks,

~ Aaron

**Comment by jjallaire** [comment]
The TensorFlow C API really only lets you feed data into and generate
predictions from models (i.e. Interact with models in a production
environment). It doesn't however let you define models, which is really the
heart of what R users want to do. It also provides no access to the higher
level TF interfaces like pre-built layers, metrics, tflearn, keras , etc.
For these things we must wrap the Python API and so we do.

The tf object exported from the package provides access to the entire API
dynamically, so doesn't need anything special to stay up to date with the
core Python API.

We also plan on providing tidy, pure functional wrappers for the higher
level TF interfaces. Hopefully will have some progress to show on this
front soon.

**Comment by jjallaire** [comment]
Note that you can technically define models using the C API however you'd
have to build them up from raw protobuf-encoded tensor operations. That
would leave aside what appears to be many thousands of person-hours that
have gone into creating the Python API's on top of raw tensors.

You could in theory do it but it would be incredibly expensive and likely
quite defect prone, whereas with wrapping the Python API you leverage all
of that existing engineering and testing rather than painstakingly
reinventing it within another language.

**Comment by pourzanj** [comment]
I'm not a contributor to Tensorflow for R, but I have been using it a lot recently, and in doing so have gotten familiar with the source code.

As of now, Tensorflow for R is indeed wrapping the Python API using an R package called [Reticulate](https://github.com/rstudio/reticulate) that allows users to use Python libraries and objects directly from R (which is pretty cool). Tensorflow's website has a sort of [guide for developer's who'd like to create Tensorflow interfaces in other languages](https://www.tensorflow.org/extend/language_bindings), e.g. R. You're right, it seems like they do prefer that new interfaces call the C API rather than the R API (to quote):

> Python was the first client language supported by TensorFlow and currently supports the most features. More and more of that functionality is being moved into the core of TensorFlow (implemented in C++) and exposed via a C API. Client languages should use the language's foreign function interface (FFI) to call into this C API to provide TensorFlow functionality.

I'm guessing they prefer this because there's less overhead than calling C than Python, but as it also says there that the Python API supports the most features and that they're still in the process of adding all that functionality to the C API. So that sort of puts R users in a tough spot for now, since we wouldn't want to lose out on certain functionality.

As an aside, if a new wrapper around the C API was written from scratch, I think it would be awesome if it could be used in the Hadley-verse style with piping and other syntactic sugar. Tensorflow is all about building complicated expression by passing Tensor through sequences of ops. That's just begging for the piping syntax us R user's know and love e.g. if we want to take the gradient of the length of a vector instead of

```r
Vec <- placeholder(float32)
SumSquaredVec <- sum(pow(Vec, 2))
LengthVec <- sqrt(SumSquareVec)
GradLengthVec <- gradients(LengthVec, Vec[1])
```

we could write

```r
GradLengthVec <- placeholder(float32) %>% pow(2) %>% sum %>% sqrt %>% gradients(Vec[1])
```

**Comment by jjallaire** [comment]
Yes, creating R APIs that have tidyverse semantics is exactly what we want
to do.

What it says on the TF website notwithstanding, wrapping the C API is a
non-starter if you want to have feature parity with the entire surface area
of TF. It would literally take tens of thousands of person hours to match
what they've done in Python. All of the higher level APIs and
utility/workflow functions are written in Python not C++.

**Comment by pourzanj** [comment]
> What it says on the TF website notwithstanding, wrapping the C API is a
non-starter if you want to have feature parity with the entire surface area
of TF. It would literally take tens of thousands of person hours to match
what they've done in Python. All of the higher level APIs and
utility/workflow functions are written in Python not C++.

Ah ok that makes sense. That makes wrapping Python the only option.

> Yes, creating R APIs that have tidyverse semantics is exactly what we want to do.

Awesome! I'd be happy to help if it means getting that functionality out sooner.

**Comment by jjallaire** [comment]
Okay, once we get the projects into a state where it's straightforward to accept contributions we'll be back in touch!

**Comment by ajschumacher** [comment]
Thanks! I updated my [post](http://planspace.org/20170320-tensorflow_apis_for_various_languages/) about the variety of TensorFlow APIs a little based on this thread. Appreciate your explanations!

**Comment by jjallaire** [comment]
Great, glad to help. Hopefully we'll have some new work on higher level APIs to share soon!
